docs: vnc selection (#24)
tari-project/rfcs: src/RFC-0314-VNCSelection.md

[BCP 14](https://tools.ietf.org/html/bcp14) (covering RFC2119 and RFC8174) when, and only when, they appear in all capitals, as 
shown here.

## Disclaimer

This document and its content are intended for information purposes only and may be subject to change or update
without notice.

This document may include preliminary concepts that may or may not be in the process of being developed by the Tari
community. The release of this document is intended solely for review and discussion by the community of the
technological merits of the potential system outlined herein.

## Goals
The goal of this RFC is to describe the process for allocating Validator Nodes (VNs) to Validator Node Committees (VNCs)

## Intro
Validator nodes will have to group themselves into Validator Node Committees for transactions/shard processing. Committees will be formed for each new transaction/shard processing. This committee needs to be determined pseudorandomly, and we use the VN_Key as this changes periodically and is pseudorandom. Each VN will be allocated a unique and individual shard space to serve in as a VNC member. 

> 

Reviewer: ```suggestion
# Process of choosing committees when an instruction needs to be processed

For each instruction, the substates involved in this instruction **MUST** be known before they can be processed. 
For each involved substate, the address of the substate is mapped to a shard (which is a specific point in the shard space). For each shard, the VN committee is constructed from `COMMITTEE_SIZE/2` VNs to the left and right of the shard, in the [VNKey Merkle tree]. 

Thus, a single instruction will have a maximum N * `COMMITTEE_SIZE` validator nodes processing it, when N is the number of involved substates.

```

Reviewer: An example here would be useful

## Committee Creation
Because we have the base layer where each VN needs to publish a registration transaction, we can get base_node and miners to keep track of all active VNs. We represent all VNs in a (balanced) Merkle tree with the VN_keys as leaves. We declare a constant `COMMITTEE_SIZE` which can be changed in the consensus constants.
For the sake of simplicity, `COMMITTEE_SIZE` MUST be an even positive integer.

> As per the Cerberus algorithm, validator nodes are responsible for managing _sub-states_, rather than contract semantics. 

Reviewer: ```suggestion
As per the Cerberus algorithm, validator nodes are responsible for managing _substates_, rather than contract semantics. 

```

A given instruction may involve dozens of sub-states, meaning that there are potentially dozens of non-overlapping committees that are required to reach a braided consensus.

> Each committee is determined independently. For each sub-state, a committee is formed by taking the first  `COMMITTEE_SIZE / 2` VN_keys to the left and `COMMITTEE_SIZE / 2` VN_Keys to the right of the sub-state's shard address in the merkle tree.

Reviewer: ```suggestion
Each committee is determined independently. For each sub-state, a committee is formed by taking the first  `Ceiling(COMMITTEE_SIZE / 2)` VN_keys to the left and `Ceiling(COMMITTEE_SIZE / 2)` VN_Keys to the right of the sub-state's shard address in the merkle tree.

```

Reviewer 2: Committee sie is even, so this doesn't matter

second layer has a delayed view of the network. VNs can also know beforehand and prep to ensure they are ready when new block heights appear. 

An important edge case here that is implied but not listed, is that if the whole network is less or equal to the `COMMITTEE_SIZE`, then the whole network participates in the VNC.

## Committee proofs
We construct the balanced Merkle tree from all active VN registration transactions and prune away all inactive ones. Because the [base node]s have to keep track
of the entire unspent UTXO set, it becomes easy for them to track and validate all active VN registration UTXOs. This means we can keep [base node]s 
responsible for validating and constructing the Merkle tree. We commit this Merkle tree per block as a Merkle root inside the block's header as the `validator_node_set_root`. 

This Merkle root in the header always lags by one block, meaning that the Merkle root is for the state of the VN's before the start of the block it's mined in. 
When a VN registers, that new VN key will only appear inside the Merkle tree in the next block. The reason for this is that header_hash is used in the
the calculation process of the VN_key. 

> ### Option 2 for Committee proofs

Reviewer: This section should be removed and merged with the above section, which needs more detail as @CjS77 suggests

The VN_key needs some random entropy that's not minable to ensure that a VN cannot choose its VN_key. This random entropy is currently the block hash of the
block the VN registration was mined in. Hence the reason for the Merkle root lagging by one block. If we change this entropy value be another source of 
randomness, such as the `utxo_merkle_root`, then we don't have to do the lag by 1, and it can all be calculated in the single block.

> ## Leader selection

Reviewer: This should be a different RFC

Reviewer 2: +1

WHETHER IN CONTRACT, STRICT LIABILITY OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

## Language

The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", 
"NOT RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in 
[BCP 14](https://tools.ietf.org/html/bcp14) (covering RFC2119 and RFC8174) when, and only when, they appear in all capitals, as 
shown here.

## Disclaimer

This document and its content are intended for information purposes only and may be subject to change or update
without notice.

This document may include preliminary concepts that may or may not be in the process of being developed by the Tari
community. The release of this document is intended solely for review and discussion by the community of the
technological merits of the potential system outlined herein.

## Goals
The goal of this RFC is to describe the process for allocating Validator Nodes (VNs) to Validator Node Committees (VNCs)

## Intro

> Validator nodes will have to group themselves into Validator Node Committees for transactions/shard processing. Committees will be formed for each new transaction/shard processing. This committee needs to be determined randomly, and we use the VN_Key as this changes periodically and is random. Each VN will be allocated a unique and individual shard space to serve in as a VNC member. 

Author: true

## Committee Creation
Because we have the base layer where each VN needs to publish a registration transaction, we can get base_node and miners to keep track of all active VNs. We represent all VNs in a (balanced) Merkle tree with the VN_keys as leaves. We declare a constant `COMMITTEE_SIZE` which can be changed in the consensus constants.
When a transaction is proposed, the shards for each substate determine the VN committee from the Merkle tree by finding the `COMMITTEE_SIZE / 2` VN_keys to the 
left and `COMMITTEE_SIZE / 2` VN_Keys to the right in the Merkle tree.

This makes it very easy to determine what shard space a VN needs to serve, which states the VN needs to sync from peers, and who has them. Because the

> second layer has a delayed view of the network. VNs can also know beforehand and prep to ensure they are ready when new block heights appear. 

Author: https://github.com/tari-project/rfcs/pull/22
Discussed in RFC-0303

## Committee proofs
We construct the balanced Merkle tree from all active VN registration transactions and prune away all inactive ones. Because the [base node]s have to keep track
of the entire unspent UTXO set, it becomes easy for them to track and validate all active VN registration UTXOs. This means we can keep [base node]s 
responsible for validating and constructing the Merkle tree. We commit this Merkle tree per block as a Merkle root inside the block's header as the `validator_node_set_root`. 

This Merkle root in the header always lags by one block, meaning that the Merkle root is for the state of the VN's before the start of the block it's mined in. 
When a VN registers, that new VN key will only appear inside the Merkle tree in the next block. The reason for this is that header_hash is used in the

> the calculation process of the VN_key. 

Reviewer 2: This needs some more background and justification.

There are a few goals that base node registration should be looking to achieve:
1. Sybil resistance
2. Providing the source of truth for the population of VNs at all times.
3. Providing the basis for the mapping of VNs to shard space
4. Managing VN registration and de-registration
5. Providing a level of stability regarding the validator node population

1, 2, 4 can be fully accomplished by using a new type of output feature marking an output as a `VN_REGISTRATION` transaction. The UTXO will lock up some non-negligible amount (`VN_DEPOSIT` of say 10,000 XTR) for a minimum amount of time (`VN_DEPOSIT_PERIOD`), say 3 months. 

The deposit and time lock act as an effective form of Sybil resistance (Goal 1). 
The set of inspect `VN_REGISTRATION` outputs represents the population of validator nodes (Goal 2).
After the minimum time period, VN operators can spend the `VN_REGISTRATION` output to deregister the node (Goal 4). 

A more aggressive alternative is to _require_ via an output script & covenant that the UTXO gets renewed or deregistered every 3 months, otherwise, the UTXO is spendable by anyone to a burn output.

Goal 3 is partially achieved via the registration transaction. However, we must shuffle VN's shard groups periodically to prevent strategies by node operators to carry out some [shard targeting attacks]*. This is done by introducing the concept of an _epoch_ into the VN population management philosophy. 

 Epochs are periods of time, of roughly hours (720 blocks), that commit to, and define the current VN set. Any transactions involving `VN_REGISTRATION` outputs do not take effect until the start of the next epoch. This lets us achieve Goal 5 directly, and provides the additional security to achieve Goal 3. There are some other nuances of epoch management that require their own RFC, but broadly speaking, the base layer PoW chain acts as a big, slow metronome keeping time on the DAN.

*explanation required

Author: Busy working on RFC_0313 which deals with VN Registration. This will address most of this. 
Do you think its perhaps better to merge the two, rather than have one for registration and one for VNC selection?

### Option 2 for Committee proofs
The VN_key needs some random entropy that's not minable to ensure that a VN cannot choose its VN_key. This random entropy is currently the block hash of the
block the VN registration was mined in. Hence the reason for the Merkle root lagging by one block. If we change this entropy value be another source of 
randomness, such as the `utxo_merkle_root`, then we don't have to do the lag by 1, and it can all be calculated in the single block.

## Leader selection
The VNC leader should not be decided beforehand and must be chosen randomly only when a tx is published. We also need to select an order of leaders

> in the case, the first leader is offline/and or not responsive.

Reviewer 2: Every instance of the word "random" in this RFC is misleading. Everything must be deterministic, otherwise, there can be no consensus.
